README
pyAggr3g470r
============

#### A simple Python news aggregator.

.. Use headers in this order #=~-_

:toc: yes
:symrefs: yes
:sortrefs: yes
:compact: yes
:subcompact: no
:rfcedstyle: no
:comments: no
:inline: yes
:private: yes

:author: Cédric Bonhomme
:contact: http://cedricbonhomme.org/


Presentation
------------
pyAggr3g470r_ is a multi-threaded news aggregator with a web interface
based on CherryPy_. Articles are stored in a MongoDB_ base.



Features
------------
* articles are stored in a MongoDB_ database;
* find an article with history;
* e-mail notification;
* export articles to HTML, EPUB, PDF or raw text;
* mark or unmark articles as favorites;
* share articles with Diaspora, Google Buzz, Pinboard, delicious, Identi.ca, Digg, reddit, Scoopeo, Blogmarks and Twitter;
* generation of QR code with the content or URL of an article. So you can read an article later on your smartphone (or share with friends).



Requierements
-------------
Software required
~~~~~~~~~~~~~~~~~
* Python_ 2.7.*;
* MongoDB_ and PyMongo_;
* feedparser_;
* CherryPy_ (version 3 and up);
* BeautifulSoup_.


Optional module
~~~~~~~~~~~~~~~
These modules are not required but enables more features:
* lxml and Genshi;
* Python Imaging Library for the generation of QR codes.


If you want to install these modules:

.. code-block:: bash

    $ sudo aptitude install  python-lxml python-genshi


Backup
------

If you want to backup your database:

.. code-block:: bash

    $ su
    $ /etc/init.d/mongodb stop
    $ cp /var/lib/mongodb/pyaggr3g470r.* ~


Donnation
---------
If you wish and if you like pyAggr3g470r, you can donate via bitcoin. My bitcoin address: 1GVmhR9fbBeEh7rP1qNq76jWArDdDQ3otZ
Thank you!



License
------------
pyAggr3g470r_ is under GPLv3_ license.


.. _Python: http://python.org/
.. _pyAggr3g470r: https://bitbucket.org/cedricbonhomme/pyaggr3g470r/
.. _feedparser: http://feedparser.org/
.. _MongoDB: http://www.mongodb.org/
.. _PyMongo: https://github.com/mongodb/mongo-python-driver
.. _CherryPy: http://cherrypy.org/
.. _BeautifulSoup: http://www.crummy.com/software/BeautifulSoup/
.. _GPLv3: http://www.gnu.org/licenses/gpl-3.0.txt


NEWS.rst
            pyAggr3g470r project news

6.6: 2015-06-02
    New:
     * it is now possible to sort the list of articles by feed title or
     article title;
     * added base unittests. 
    Improvements:
     * fixed some minor bugs;
     * improved the asyncio crawler;
     * automatically use the good Python executable for the asyncio crawler;
     * improved controllers (enforcing better use of user_id);
     * the search is now case insensitive.

6.5.5: 2015-04-22
    The full text search powered by Whoosh has been removed.

6.5.4: 2015-04-16
    This release introduces a new config module and a new search functionality.
    The result of a search is integrated in the main page.

6.5.3: 2015-04-14
    The fetch call is now dependent to the selected crawling method.

6.5.2: 2015-04-14
    The look and feel has been globally improved.
    It is now possible to add a new feed from any page via a dropdown menu.

6.5.1: 2015-04-08
    Improvements:
     * improvements on the controllers;
     * the minimum error count is now specified in the configuration file.

6.5: 2015-04-07 
    Improvements:
     * new CSS;
     * improved installation script;
     * it is now possible to delete all duplicate articles with one HTTP
     delete request.

6.4: 2015-03-17
    New:
     * a new page 'history' to explore your database of articles.
    Changes:
     * updated documentation;
     * minor improvements.
    Fixes:
     * changed the type of the column 'last_modified' to string.

6.3: 2015-03-08
    New:
    * a new architecture with base for controllers;
    * new, more complete RESTful API;
    * a crawler handling errors and cache control;
    * the new crawler takes advantage of the RESTful API
    (can be run on the client side).

6.2: 2015-02-26
    The system of email notifications for new articles has been removed.
    This feature was hardly used.

6.1: 2015-02-23
    Improvements: articles are now identified with the id provided
    by the RSS/ATOM feed.
    Prevent BeautifulSoup4 from adding extra '<html><body>' tags to
    the soup with the 'lxml' parser.
    Indexation is now used with the new crawler.
    The documentation has been updated.

6.0: 2015-02-08
    New: pyAggr3g470r is now working with Python 3.4. Tested on Heroku
    with Python 3.4.2.
    Improvements: The feed crawler uses the PEP 3156 (asyncio). The action
    buttons are now on the left. It is easier to mark an article as read.

5.7.0: 2014-11-20
    Improvements: major improvements of the email notifications system.
    Notifications are now sent through Postmark (for example for Heroku)
    or a standard SMTP server.
    Bug fix: activation key was too long for the database column.

5.6.6: 2014-09-24
    Improvements: Significant performance improvement for the views
    /favorites and /unread. The performance problem has been introduced
    with the release 5.6.5.

5.6.5: 2014-09-15
    This release fixes a major bug introduced with the version 0.9.7 of SQLAlchemy
    (conflicts between persistent ant transcient SQLAlchemy objects).

5.6.4: 2014-09-12
    Translations have been updated.
    Gravatar extension has been removed.
    Minor fix.

5.6.3: 2014-08-27
    News: It is now possible to delete articles older than a given number
    of weeks. This can be done via the management page.
    A new environment variable enables to activate or deactivate the
    self-registration.
    Changes: translations were updated. Some minor bugfixes. Improved
    deployment instructions in the README.

5.6.2: 2014-08-10
    Minor improvements: Articles are automatically retrieved after the import
    of an OPML file.
    When displaying all articles (unread + read), titles of unread articles
    are emphasized in bold.
    A new tab is opened when clicking on the title of an article.
    New: pyAggr3g470r can be deployed with the Heroku auto deploy button.

5.6.1: 2014-07-13
    Performance improvements: faster database insertion of articles by
    the crawler and loading of the management page.
    Minor bug fixes.

5.6: 2014-07-05
    pyAggr3g470r has now a RESTful JSON API which enables to manage Feed and
    Article objects. The API is documented in the README file.
    The main page is using a subset of this API with jQuery HTTP requests.

5.5: 14 Jun 2014 13:09
    This release introduces a redesigned homepage which loads much faster and
    is easier to read. pyAggr3g470r can now be run by Apache.
    Adding a feed no longer requires a title and a site link.

5.4: 28 May 2014 05:16
    This version makes it possible for a user to export all of their feeds and
    articles as a JSON file for later restoration.

5.3: 23 May 2014 22:39
    This release introduces some UI improvements, especially for the home page.

5.2: 16 May 2014 17:23
    This release adds minor bug fixes and UI improvements.

5.1: 13 May 2014 20:58
    When deployed on Heroku, the platform now uses the Postmark service to
    send account confirmation emails to users. It is no longer required to
    enter a first name and a last name to create an account.

5.0: 04 May 2014 20:54
    pyAggr3g470r is now translated into English and French. Improvements
    concerning the news retriever and the Whoosh search functionality have
    been made. The user can now export all articles in JSON format.
    The user of the platform now has the possibility to delete his or her
    account.

4.9: 24 Apr 2014 21:11
    This version introduces minor improvements to the Jinja templates and
    bugfixes (relating to the import of OPML files with sub-categories and
    relating to the Whoosh index generation).

4.8: 13 Apr 2014 11:35
    Feeds are now retrieved in a separated process with the Gevent library.
    This offers the best performance on Heroku.

4.7: 12 Apr 2014 12:24
    pyAggr3g470r can now be deployed on Heroku or on a traditional server.
    Moreover, several users can use an instance of pyAggr3g470r. A platform is

utils.py
#! /usr/bin/env python
#-*- coding: utf-8 -*-

# pyAggr3g470r - A Web based news aggregator.
# Copyright (C) 2010  Cédric Bonhomme - http://cedricbonhomme.org/
#
# For more information : http://bitbucket.org/cedricbonhomme/pyaggr3g470r/
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>

__author__ = "Cedric Bonhomme"
__version__ = "$Revision: 1.2 $"
__date__ = "$Date: 2010/12/07 $"
__revision__ = "$Date: 2011/04/15 $"
__copyright__ = "Copyright (c) Cedric Bonhomme"
__license__ = "GPLv3"

#
# This file provides functions used for:
# - the database management;
# - generation of tags cloud;
# - HTML processing;
# - mail notifications.
#

import re
import hashlib
import sqlite3
import operator
import urlparse
import calendar
import unicodedata
import htmlentitydefs

import articles

import smtplib
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText

import urllib2
import BaseHTTPServer
from BeautifulSoup import BeautifulSoup

from datetime import datetime
from string import punctuation
from collections import Counter
from collections import OrderedDict

from StringIO import StringIO

import threading
LOCKER = threading.Lock()

import os
import ConfigParser
# load the configuration
config = ConfigParser.RawConfigParser()
try:
    config.read("./cfg/pyAggr3g470r.cfg")
except:
    config.read("./cfg/pyAggr3g470r.cfg-sample")
path = os.path.abspath(".")

MONGODB_ADDRESS = config.get('MongoDB', 'address')
MONGODB_PORT = int(config.get('MongoDB', 'port'))
MONGODB_USER = config.get('MongoDB', 'user')
MONGODB_PASSWORD = config.get('MongoDB', 'password')

MAX_NB_ARTICLES = int(config.get('global', 'max_nb_articles'))

mail_from = config.get('mail','mail_from')
mail_to = config.get('mail','mail_to')
smtp_server = config.get('mail','smtp')
username =  config.get('mail','username')
password =  config.get('mail','password')

DIASPORA_POD = config.get('misc', 'diaspora_pod')

# regular expression to chech URL
url_finders = [ \
    re.compile("([0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}|(((news|telnet|nttp|file|http|ftp|https)://)|(www|ftp)[-A-Za-z0-9]*\\.)[-A-Za-z0-9\\.]+)(:[0-9]*)?/[-A-Za-z0-9_\\$\\.\\+\\!\\*\\(\\),;:@&=\\?/~\\#\\%]*[^]'\\.}>\\),\\\"]"), \
    re.compile("([0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}|(((news|telnet|nttp|file|http|ftp|https)://)|(www|ftp)[-A-Za-z0-9]*\\.)[-A-Za-z0-9\\.]+)(:[0-9]*)?"), \
    re.compile("(~/|/|\\./)([-A-Za-z0-9_\\$\\.\\+\\!\\*\\(\\),;:@&=\\?/~\\#\\%]|\\\\)+"), \
    re.compile("'\\<((mailto:)|)[-A-Za-z0-9\\.]+@[-A-Za-z0-9\\.]+"), \
]

def detect_url_errors(list_of_urls):
    """
    Detect URL errors.
    Return a list of error(s).
    """
    errors = []
    for url in list_of_urls:
        req = urllib2.Request(url)
        try:
            urllib2.urlopen(req)
        except urllib2.HTTPError, e:
            # server couldn't fulfill the request
           errors.append((url, e.code, \
                BaseHTTPServer.BaseHTTPRequestHandler.responses[e.code][1]))
        except urllib2.URLError, e:
            # failed to reach the server
            errors.append((url, e.reason.errno ,e.reason.strerror))
    return errors

def clear_string(data):
    """
    Clear a string by removing HTML tags, HTML special caracters
    and consecutive white spaces (more that one).
    """
    p = re.compile(r'<[^<]*?/?>') # HTML tags
    q = re.compile(r'\s') # consecutive white spaces
    return p.sub('', q.sub(' ', data))

def unescape(text):
    """
    Removes HTML or XML character references and entities from a text string.
    """
    def fixup(m):
        text = m.group(0)
        if text[:2] == "&#":
            # character reference
            try:
                if text[:3] == "&#x":
                    return unichr(int(text[3:-1], 16))
                else:
                    return unichr(int(text[2:-1]))
            except ValueError:
                pass
        else:
            # named entity
            try:
                text = unichr(htmlentitydefs.name2codepoint[text[1:-1]])
            except KeyError:
                pass
        return text # leave as is
    return re.sub("&#?\w+;", fixup, text)

def not_combining(char):
    return unicodedata.category(char) != 'Mn'

def strip_accents(text, encoding):
    """
    Strip accents.

    >>> print strip_accents("déjà", "utf-8")
    deja
    """
    unicode_text= unicodedata.normalize('NFD', text.decode(encoding))
    return filter(not_combining, unicode_text).encode(encoding)

def normalize_filename(name):
    """
    Normalize a file name.
    """
    file_name = re.sub("[,'!?|&]", "", name)
    file_name = re.sub("[\s.]", "_", file_name)
    file_name = file_name.strip('_')
    file_name = file_name.strip('.')
    file_name = strip_accents(file_name, "utf-8")
    return os.path.normpath(file_name)

def top_words(articles, n=10, size=5):
    """
    Return the n most frequent words in a list.
    """
    words = Counter()
    wordre = re.compile(r'\b\w{%s,}\b' % size, re.I)
    for article in articles:
        for word in wordre.findall(clear_string(article["article_content"])):
            words[word.lower()] += 1
    return words.most_common(n)

def tag_cloud(tags, query="word_count"):
    """
    Generates a tags cloud.
    """
    tags.sort(key=operator.itemgetter(0))
    if query == "word_count":
        # tags cloud from the management page
        return ' '.join([('<font size=%d><a href="/search/?query=%s" title="Count: %s">%s</a></font>\n' % \
                    (min(1 + count * 7 / max([tag[1] for tag in tags]), 7), word, count, word)) \
                        for (word, count) in tags])
    if query == "year":
        # tags cloud for the history
        return ' '.join([('<font size=%d><a href="/history/?query=%s:%s" title="Count: %s">%s</a></font>\n' % \
                        (min(1 + count * 7 / max([tag[1] for tag in tags]), 7), query, word, count, word)) \
                            for (word, count) in tags])
    return ' '.join([('<font size=%d><a href="/history/?query=%s:%s" title="Count: %s">%s</a></font>\n' % \
                        (min(1 + count * 7 / max([tag[1] for tag in tags]), 7), query, word, count, calendar.month_name[int(word)])) \
                            for (word, count) in tags])

def send_mail(mfrom, mto, feed_title, article_title, description):
    """
    Send the article via mail.
    """
    # Create the body of the message (a plain-text and an HTML version).
    html = """<html>\n<head>\n<title>%s</title>\n</head>\n<body>\n%s\n</body>\n</html>""" % \
                (feed_title + ": " + article_title, description)
    text = clear_string(description)

    # Create message container - the correct MIME type is multipart/alternative.
    msg = MIMEMultipart('alternative')
    msg['Subject'] = '[pyAggr3g470r] ' + feed_title + ": " + article_title
    msg['From'] = mfrom
    msg['To'] = mto

    # Record the MIME types of both parts - text/plain and text/html.
    part1 = MIMEText(text, 'plain')
    part2 = MIMEText(html, 'html')

    # Attach parts into message container.
    # According to RFC 2046, the last part of a multipart message, in this case
    # the HTML message, is best and preferred.
    msg.attach(part1)
    msg.attach(part2)

    # Send the message via local SMTP server.
    s = smtplib.SMTP(smtp_server)
    s.login(username, password)
    # sendmail function takes 3 arguments: sender's address, recipient's address
    # and message to send - here it is sent as one string.
    s.sendmail(mfrom, mto, msg.as_string())
    s.quit()

def string_to_datetime(stringtime):
    """
    Convert a string to datetime.
    """
    date, time = stringtime.split(' ')
    year, month, day = date.split('-')
    hour, minute, second = time.split(':')
    return datetime(year=int(year), month=int(month), day=int(day), \
        hour=int(hour), minute=int(minute), second=int(second))

def compare(stringtime1, stringtime2):
    """
    Compare two dates in the format 'yyyy-mm-dd hh:mm:ss'.
    """
    datetime1 = string_to_datetime(stringtime1)
    datetime2 = string_to_datetime(stringtime2)
    if datetime1 < datetime2:
        return -1
    elif datetime1 > datetime2:
        return 1
    return 0

def add_feed(feed_url):
    """
    Add the URL feed_url in the file feed.lst.
    """
    if os.path.exists("./var/feed.lst"):
        for line in open("./var/feed.lst", "r"):
            if feed_url in line:
                # if the feed is already in the file
                return False
    with open("./var/feed.lst", "a") as f:
        f.write(feed_url + "\n")
    return True

def change_feed_url(old_feed_url, new_feed_url):
    """
    Change the URL of a feed given in parameter.
    """
    # Replace the URL in the text file
    with open("./var/feed.lst", "r") as f:
        lines = f.readlines()
    try:
        lines[lines.index(old_feed_url+"\n")] = new_feed_url + '\n'
    except:
        return
    with open("./var/feed.lst", "w") as f:
        f.write("\n".join(lines))

    # Replace the URL in the data base.
    try:
        conn = sqlite3.connect(sqlite_base, isolation_level = None)
        c = conn.cursor()
        c.execute("UPDATE articles SET feed_link='" + new_feed_url + "' WHERE feed_link='" + old_feed_url +"'")
        c.execute("UPDATE feeds SET feed_link='" + new_feed_url + "' WHERE feed_link='" + old_feed_url +"'")
        conn.commit()
        c.close()
    except Exception, e:
        print e

def change_feed_name(feed_url, new_feed_name):
    """
    Change the name of a feed given in parameter.
    """
    try:
        conn = sqlite3.connect(sqlite_base, isolation_level = None)
        c = conn.cursor()
        c.execute('UPDATE feeds SET feed_title="' + new_feed_name + '" WHERE feed_link="' + feed_url +'"')
        conn.commit()
        c.close()
    except Exception, e:
        print e

def change_feed_logo(feed_url, new_feed_logo):
    """
    Change the logo of a feed given in parameter.
    """
    try:
        conn = sqlite3.connect(sqlite_base, isolation_level = None)
        c = conn.cursor()
        c.execute('UPDATE feeds SET feed_image_link="' + new_feed_logo + '" WHERE feed_link="' + feed_url +'"')
        conn.commit()
        c.close()
    except Exception, e:
        print e

def remove_feed(feed_url):
    """
    Remove a feed from the file feed.lst and from the SQLite base.
    """
    feeds = []
    # Remove the URL from the file feed.lst
    if os.path.exists("./var/feed.lst"):
        for line in open("./var/feed.lst", "r"):
            if feed_url not in line:
                feeds.append(line.replace("\n", ""))
        with open("./var/feed.lst", "w") as f:
            f.write("\n".join(feeds) + "\n")

def search_feed(url):
    """
    Search a feed in a HTML page.
    """
    soup = None
    try:
        page = urllib2.urlopen(url)
        soup = BeautifulSoup(page)
    except:
        return None
    feed_links = soup('link', type='application/atom+xml')
    feed_links.extend(soup('link', type='application/rss+xml'))
    for feed_link in feed_links:
        if url not in feed_link['href']:
            return urlparse.urljoin(url, feed_link['href'])
        return feed_link['href']
    return None